File: 2016/Day9/Day9.pl
use strict;
use warnings;

sub DecompressV1 {
  my $compressedtext = shift;

  my $pos = index($compressedtext, '(');
  if ($pos == -1) {
    return length($compressedtext);
  } elsif ($compressedtext =~ m/(\((\d+)x(\d+)\))/) {
    my ($bracket, $size, $repeat) = ($1, $2, $3);
    my $i = $pos + length($bracket);
    my $remainingText = substr($compressedtext, $i + $size);
    return $pos + ($size * $repeat) + DecompressV1($remainingText);
  } else {
    return 0;
  }
}

sub DecompressV2 {
  my $compressedtext = shift;

  my $pos = index($compressedtext, '(');
  if ($pos == -1) {
    return length($compressedtext);
  } elsif ($compressedtext =~ m/(\((\d+)x(\d+)\))/) {
    my ($bracket, $size, $repeat) = ($1, $2, $3);
    my $i = $pos + length($bracket);
    my $textToRepeat = substr($compressedtext, $i, $size);
    my $remainingText = substr($compressedtext, $i + $size);
    return $pos + (DecompressV2($textToRepeat) * $repeat) + DecompressV2($remainingText);
  } else {
    return 0;
  }
}

my $inputfile = shift;
if (!defined($inputfile)) {
  $inputfile = 'sample1.txt';
}

open(my $fh, '<', $inputfile) or die "Couldn't open $inputfile: $!";
chomp(my @lines = <$fh>);
close($fh);

my $compressedtext = '';
foreach my $line (@lines) {
  $line =~ s/(?:\R|\s)//g; # Trim newlines and whitespace
  $compressedtext = $compressedtext . $line;
}

my $len = DecompressV1($compressedtext);
print "V1 Length: $len\n";

$len = DecompressV2($compressedtext);
print "V2 Length: $len\n";

File: 2016/Day9/Day9.cpp
// Day9.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"
#include <fstream>
#include <iostream>
#include <regex>
#include <string>
#include <vector>

std::wstring GetCompressedString(const std::wstring& inputFile) {
   std::wstring str;
   std::wifstream input{ inputFile };
   if (!input.is_open()) {
      return str;
   }

   do {
      std::wstring line;
      std::getline(input, line);
      str.append(line);
   } while (input.good());

   return str;
}

template <int> size_t Decompress(const std::wstring& input);

template <> size_t Decompress<1>(const std::wstring& input) {
   size_t pos{ input.find(L'(') };
   if (pos == std::wstring::npos) {
      return input.length();
   }
   else {
      std::wregex re{ L"\\((\\d+)x(\\d+)\\)" };
      std::wsmatch m;

      if (!std::regex_search(input, m, re)) {
         return 0;
      }

      std::wstring bracket{ m[0] };
      size_t size{ static_cast<size_t>(_wtoi(m[1].str().c_str())) }, repeat{ static_cast<size_t>(_wtoi(m[2].str().c_str())) };
      size_t index{ pos + bracket.length() };
      return pos + (size * repeat) + Decompress<1>(input.substr(index + size));
   }
}

template <> size_t Decompress<2>(const std::wstring& input) {
   size_t pos{ input.find(L'(') };
   if (pos == std::wstring::npos) {
      return input.length();
   }
   else {
      std::wregex re{ L"\\((\\d+)x(\\d+)\\)" };
      std::wsmatch m;

      if (!std::regex_search(input, m, re)) {
         return 0;
      }

      std::wstring bracket{ m[0] };
      size_t size{ static_cast<size_t>(_wtoi(m[1].str().c_str())) }, repeat{ static_cast<size_t>(_wtoi(m[2].str().c_str())) };
      size_t index{ pos + bracket.length() };
      return pos + (Decompress<2>(input.substr(index, size)) * repeat) + Decompress<2>(input.substr(index + size));
   }
}

int wmain(int argc, wchar_t *argv[]) {
   std::wstring inputFile;
   if (argc == 2) {
      inputFile = argv[1];
   }
   else {
      inputFile = L"sample2.txt";
   }

   std::wstring compressedText{ GetCompressedString(inputFile) };

   std::wcout << L"First: " << Decompress<1>(compressedText) << std::endl;
   std::wcout << L"Second: " << Decompress<2>(compressedText) << std::endl;

   return 0;
}
